### DISCUSSION TITLE
Right way to link multilple OAuth to profile

### DISCUSSION THREAD

**Discussion by Kam1k4dze** [pocketbase/pocketbase#4712] [Q&A] [answered]
I would like to link multiple external authorizations to one profile.
Currently, I log into the profile and call `await pb.collection('users').authWithOAuth2({ provider });`, which works.
However, I am encountering issues when another profile with the same OAuth already exists, as it signs me into that one instead.

Additionally, what is the recommended method for deleting an account that does not have any connected OAuth2 accounts?

**Comment by ganigeorgiev** [comment, marked as answer]
> However, I am encountering issues when another profile with the same OAuth already exists, as it signs me into that one instead.

We don't support "shared" OAuth2 accounts, if that's what you are asking. 1 external provider can be linked to only 1 PocketBase auth collection record. I don't have context about your use case and what you are trying to do but if you want to return an error instead of authenticating with the matched user you can attach to the [`OnRecordBeforeAuthWithOAuth2Request` hook](https://pocketbase.io/docs/go-event-hooks/#onrecordbeforeauthwithoauth2request) and compare the event record (if any) with the auth state from the request context. Here is one possible example using the JS pb_hooks (_I haven't tested it_):

```js
onRecordBeforeAuthWithOAuth2Request((e) => {
    // https://pocketbase.io/docs/js-routing/#retrieving-the-current-auth-state
    const authRecord = e.httpContext.get("authRecord")
    if (authRecord && e.record && e.record.id != authRecord.id) {
        throw new BadRequestError("The authenticated user that sent the request is different from the OAuth2 linked one")
    }
}, "users")
```


> what is the recommended method for deleting an account that does not have any connected OAuth2 accounts?

I'm not sure I follow. In general calling `pb.collection("users").delete("userId")` should be enough to delete a user (any linked external auth providers should be automatically cascade deleted). If you want to check whether the user has linked OAuth2 accounts before performing the delete you can call the [`listExternalAuths` endpoint](https://pocketbase.io/docs/api-records/#list-linked-external-auth-providers) (_or perform the check programmatically via a before delete hook_).

**Reply by Kam1k4dze** [reply]
Thank you so much! This is exactly what I needed. Sorry for the unclear wording. To use with Go, the code should look like this?
```go
app.OnRecordBeforeAuthWithOAuth2Request("users").Add(func(e *core.RecordAuthWithOAuth2Event) error {
		authRecord := e.HttpContext.Get("authRecord")
		if (authRecord != nil) && (e.Record != nil) && (authRecord != e.Record.Id) {
			return apis.NewBadRequestError("The authenticated user that sent the request is different from the OAuth2 linked one", "")
		}
		return nil
	})
```


I meant to automatically delete the profile when the last OAuth2 is unlinked. I'm thinking to use `OnRecordAfterUnlinkExternalAuthRequest` hook and check if there are any external oauth providers. I need advice on how to get the list of external providers using Go API

**Reply by ganigeorgiev** [reply]
> To use with Go, the code should look like this?

Almost. Here is a Go version of the above code:

```go
app.OnRecordBeforeAuthWithOAuth2Request("users").Add(func(e *core.RecordAuthWithOAuth2Event) error {
    authRecord, _ := e.HttpContext.Get(apis.ContextAuthRecordKey).(*models.Record)
    if authRecord != nil && e.Record != nil && e.Record.Id != authRecord.Id {
        return apis.NewBadRequestError("The authenticated user that sent the request is different from the OAuth2 linked one", nil)
    }
    return nil
})
```

> I meant to automatically delete the profile when the last OAuth2 is unlinked. 

One possible sample code could be (_I haven't tested it_):

```go
app.OnRecordAfterUnlinkExternalAuthRequest("users").Add(func(e *core.RecordUnlinkExternalAuthEvent) error {
    externalAuths, err := app.Dao().FindAllExternalAuthsByRecord(e.Record)
    if err != nil {
        // return the error or just log since this is happening in an after hook
        app.Logger().Error("failed to fetch other external auths", "error", err, "record", e.Record.Id)
    } else if len(externalAuths) == 0 {
        // no other external auths -> delete automatically the user record
        if err := app.Dao().DeleteRecord(e.Record); err != nil {
            // return the error or just log since this is happening in an after hook
            app.Logger().Error("failed to automatically delete non-linked user", "error", err, "record", e.Record.Id)
        }
    }

    return nil
})
```
